sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala

```diff
@@ -106,15 +106,16 @@ class FileScanRDD(
         // don't need to run this `if` for every record.
         val preNumRecordsRead = inputMetrics.recordsRead
         if (nextElement.isInstanceOf[ColumnarBatch]) {
+          incTaskInputMetricsBytesRead()
```

Comment:
I see, so always update when processing `ColumnarBatch`, but use the previous logic otherwise. That seems OK. It should still address the original problem.

Reply:
... I guess the only possible drawback is that if the number of records in a ColumnarBatch is pretty small, then this could cause it to update bytes read a lot more frequently than before. Bu if the number of records is large (>100) then this won't matter.

Reply:
4096 is the default number of the batch reader in both ORC and Parquet. If the users set the conf to a much smaller number, they will face the perf regression due to the the extra overhead in many places. I do not think end users will do this.

Reply:
Makes sense. In this case the behavior should be the same before and after this change, but it's therefore fine, too.

Reply:
Considering that the default value of `"spark.sql.parquet.columnarReaderBatchSize` is 4096, this change is better .